File: Printly-backend/app/schemas/book.py
from pydantic import BaseModel, Field, ConfigDict
from uuid import UUID
from datetime import datetime
from decimal import Decimal


class BookMaterialItem(BaseModel):
    material_id: UUID
    quantity_per_copy: Decimal = Field(..., gt=0)


class BookCreate(BaseModel):
    title: str = Field(..., max_length=300)
    subject: str | None = Field(None, max_length=200)
    total_pages: int = Field(..., gt=0)
    color_mode: str = Field("bw", pattern="^(bw|color)$")
    sides_per_page: int = Field(1, ge=1, le=4)
    copies: int = Field(1, gt=0)
    binding_type: str | None = Field(None, max_length=50)
    has_lamination: bool = False
    notes: str | None = None
    materials: list[BookMaterialItem] = Field(default_factory=list)


class BookUpdate(BaseModel):
    title: str | None = Field(None, max_length=300)
    subject: str | None = Field(None, max_length=200)
    total_pages: int | None = Field(None, gt=0)
    color_mode: str | None = Field(None, pattern="^(bw|color)$")
    sides_per_page: int | None = Field(None, ge=1, le=4)
    copies: int | None = Field(None, gt=0)
    binding_type: str | None = Field(None, max_length=50)
    has_lamination: bool | None = None
    notes: str | None = None


class BookMaterialResponse(BaseModel):
    material_id: UUID
    material_name: str
    quantity_per_copy: Decimal
    price_per_unit: Decimal

    model_config = ConfigDict(from_attributes=True)


class BookResponse(BaseModel):
    id: UUID
    title: str
    subject: str | None
    total_pages: int
    color_mode: str
    sides_per_page: int
    copies: int
    binding_type: str | None
    has_lamination: bool
    notes: str | None
    file_size: int | None
    file_url: str | None
    created_at: datetime
    updated_at: datetime
    book_materials: list[BookMaterialResponse] = []

    model_config = ConfigDict(from_attributes=True)


class BooksRequest(BaseModel):
    title: str | None = None
    subject: str | None = None
    offset: int = 0
    limit: int = 10
    order_by: str = "created_at"
    order_dir: str = "desc"
    has_file: bool | None = None


class BookListResponse(BaseModel):
    total: int
    items: list[BookResponse]

File: Printly-frontend/src/types/book.ts
// ==================== Book Types ====================

export interface BookMaterialItem {
	material_id: string;
	quantity_per_copy: number;
}

export interface BookMaterialResponse {
	material_id: string;
	material_name: string;
	quantity_per_copy: number;
	price_per_unit: number;
}

export interface BookResponse {
	id: string;
	title: string;
	subject: string | null;
	total_pages: number;
	color_mode: string;
	sides_per_page: number;
	copies: number;
	binding_type: string | null;
	has_lamination: boolean;
	notes: string | null;
	file_size: number | null;
	file_url: string | null;
	created_at: string;
	updated_at: string;
	book_materials: BookMaterialResponse[];
}

export interface BookCreate {
	title: string;
	subject?: string;
	total_pages: number;
	color_mode?: string;
	sides_per_page?: number;
	copies?: number;
	binding_type?: string;
	has_lamination?: boolean;
	notes?: string;
	materials?: BookMaterialItem[];
	file?: File;
}

export interface BookUpdate {
	title?: string;
	subject?: string;
	total_pages?: number;
	color_mode?: string;
	sides_per_page?: number;
	copies?: number;
	binding_type?: string;
	has_lamination?: boolean;
	notes?: string;
}

export interface BooksRequest {
	title?: string;
	subject?: string;
	has_file?: boolean;
	offset?: number;
	limit?: number;
	order_by?: string;
	order_dir?: string;
}

export interface BookListResponse {
	total: number;
	items: BookResponse[];
}
